router/substituters.go

```diff
@@ -0,0 +1,73 @@
+package router
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+)
+
+var envvarTokens = regexp.MustCompile(`\$\{.+?\}`)
+var fieldTokens = regexp.MustCompile(`%\{.+?\}`)
+
+// substitute performs a key-value substitution on `obj`, replacing instances
+// of tokenMatcher in the keys or values of `obj` with `replacer(name)`. It
+// returns the substituted map and does not modify the input map.
+func substitute(
+	obj map[string]interface{}, tokenMatcher *regexp.Regexp, replacer func(key string) string,
+) map[string]interface{} {
+	newObj := make(map[string]interface{})
+	for k, v := range obj {
+		switch v := v.(type) {
+		case string:
+			newObj[k] = tokenMatcher.ReplaceAllStringFunc(v, replacer)
+		case []string:
+			newV := make([]string, len(v))
+			for i := 0; i < len(v); i++ {
+				newV[i] = tokenMatcher.ReplaceAllStringFunc(v[i], replacer)
+			}
+			newObj[k] = newV
+		default:
+			newObj[k] = v
+		}
+	}
+	return newObj
+}
+
+func substituteEnvVars(data map[string]interface{}) (map[string]interface{}, error) {
```

Comment:
Can you add doc strings similar to `substitute` for this function and `substituteFields`?